Allow Water97FluidProperties to work with a conservative variable set

lindsayad opened the pull request:
Makes heavy use of inversion via local Newton solves. Also introduces the pretty awesome compound AD type concept into `SinglePhaseFluidProperties`. In my opinion this method could make hand-coding of the 5 parameter routines (with dz_dx, dz_dy) obsolete or at least it's a very viable alternative if people are worried about computing dz_dx and dz_dy correctly by hand.

Closes #22738

lindsayad committed:
AD-ify Water97FluidProperties and support compressible discretizations

lindsayad committed:
Create assignQoIs and xyDerivatives base routines

lindsayad committed:
Use xyDerivatives

lindsayad committed:
FP Jacobian checks passing; no exceptions; system Jac bad

I don't know why it's bad ... maybe some other fluid property
derivatives I haven't AD'ified?

lindsayad committed:
Change ifdef DEBUG to ifndef NDEBUG

lindsayad committed:
FP-ADify h_from_p_T

lindsayad committed:
Registering boss Jacobians now

Closes #22738

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1238784/) on ede5fa8 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22739/site)

This comment will be updated on new commits.

lindsayad committed:
Consolidate 2D Newton inversion code

lindsayad committed:
Try to appease clang compilers

lindsayad commented:
@GiudGiud the bicubic tabulated fluid properties tests seem pretty poor. I assumed that I must have done something really bad to make those conversions fail but the behavior on `devel` seems worse!
```
*** Warning ***
The following warning occurred in the object "tabulated", of type "TabulatedBicubicFluidProperties".

Conversion from (v, e)=(0.0137352, -93378.1) to (p, T) failed


*** Warning ***
The following warning occurred in the object "tabulated", of type "TabulatedBicubicFluidProperties".

While creating (p,T) from (v,e) interpolation tables,
Inversion to (p,T) from (v,e) failed 3647 times
- 3158 of 10000 pressure values were out of user defined bounds
- 1107 of 10000 temperature values were out of user defined bounds
The pressure and temperature values were replaced with their respective user-defined min and max values.



*** Warning ***
The following warning occurred in the object "tabulated", of type "TabulatedBicubicFluidProperties".

Conversion from (v, h)=(0.0137352, -51444.1) to (p, T) failed


*** Warning ***
The following warning occurred in the object "tabulated", of type "TabulatedBicubicFluidProperties".

While creating (p,T) from (v,h) interpolation tables,
Inversion to (p,T) from (v,h) failed 3755 times
- 3223 of 10000 pressure values were out of user defined bounds
- 1238 of 10000 temperature values were out of user defined bounds
The pressure and temperature values were replaced with their respective user-defined min and max values.
```
Yea I'm not surprised that a conversion from a negative specific internal energy would fail

lindsayad commented:
I'm going to have to dig into this to know which behavior is more correct. After my PR:
```
*** Warning ***
The following warning occurred in the object "tabulated", of type "TabulatedBicubicFluidProperties".

While creating (p,T) from (v,e) interpolation tables,
Inversion to (p,T) from (v,e) failed 3647 times
- 2 of 10000 pressure values were out of user defined bounds
- 2 of 10000 temperature values were out of user defined bounds
The pressure and temperature values were replaced with their respective user-defined min and max values.



*** Warning ***
The following warning occurred in the object "tabulated", of type "TabulatedBicubicFluidProperties".

While creating (p,T) from (v,h) interpolation tables,
Inversion to (p,T) from (v,h) failed 3753 times
- 1 of 10000 pressure values were out of user defined bounds
- 2 of 10000 temperature values were out of user defined bounds
The pressure and temperature values were replaced with their respective user-defined min and max values.
```
Before my PR
```
*** Warning ***
The following warning occurred in the object "tabulated", of type "TabulatedBicubicFluidProperties".

Conversion from (v, e)=(0.0137352, -93378.1) to (p, T) failed


*** Warning ***
The following warning occurred in the object "tabulated", of type "TabulatedBicubicFluidProperties".

While creating (p,T) from (v,e) interpolation tables,
Inversion to (p,T) from (v,e) failed 3647 times
- 3158 of 10000 pressure values were out of user defined bounds
- 1107 of 10000 temperature values were out of user defined bounds
The pressure and temperature values were replaced with their respective user-defined min and max values.



*** Warning ***
The following warning occurred in the object "tabulated", of type "TabulatedBicubicFluidProperties".

Conversion from (v, h)=(0.0137352, -51444.1) to (p, T) failed


*** Warning ***
The following warning occurred in the object "tabulated", of type "TabulatedBicubicFluidProperties".

While creating (p,T) from (v,h) interpolation tables,
Inversion to (p,T) from (v,h) failed 3755 times
- 3223 of 10000 pressure values were out of user defined bounds
- 1238 of 10000 temperature values were out of user defined bounds
The pressure and temperature values were replaced with their respective user-defined min and max values.
```
Inversions failed about the same number of times but the resulting pressure and temperature values that were out of bounds are way less

lindsayad commented:
@GiudGiud do you think it makes sense to have a test in which almost a third of inversions fail? That seems like something to unit test, not something to have as an integration test. I suspect you didn't make these tests and your interns did, but I'm going to ping you since `git blame` raises your name

GiudGiud commented:
Please read the documentation for the object. 
we cannot succeed at inverting to pressure temperature if the starting point is not physical, and a good chunk of a regular grid in specific volume / energy is not physical

GiudGiud commented:
The out of bound values were caught and moved to the bounds, which would at least lead to a somewhat predictable behavior (flat interpolation). What is the new behavior?

lindsayad commented:
I’ll investigate more. Marking as draft for now

GiudGiud commented:
It seems odd that we try to invert negative energies and enthalpies though. 
I d double check those grids to see why one of the bounds is negative

GiudGiud commented:
Are you reading from a file? We could have read bad values from there.

joshuahansel commented on modules/fluid_properties/include/userobjects/Water97FluidProperties.h:
```diff
@@ -1332,6 +1453,1159 @@ class Water97FluidProperties : public SinglePhaseFluidProperties
+   * member of the pair) with respect to temperature as a function of pressure and density
+   */
+  template <typename T>
+  std::pair<T, T> T_from_p_rho(const T & p, const T & rho) const;
```
Maybe `T_drhodT_from_p_rho` would be a better name.

joshuahansel reviewed (changes requested):
I'll review more after it's not draft.

lindsayad commented:
> It seems odd that we try to invert negative energies and enthalpies though. I d double check those grids to see why one of the bounds is negative

The csv file has negative internal energy so that's what being written out by @cpgr's original test. I guess since internal energy includes a potential which you can kind of scale arbitrarily, it can be negative ... ?

lindsayad commented on modules/fluid_properties/include/utils/NewtonInversion.h:
```diff
@@ -23,40 +24,181 @@ namespace FluidPropertiesUtils
+  { return MetaPhysicL::raw_value(std::abs(f)) < tolerance; };
+  std::function<bool(const T &, const T &)> rel_tol_check = [tolerance](const T & f, const T & y)
+  { return MetaPhysicL::raw_value(std::abs(f / y)) < tolerance; };
+  auto convergence_check = MetaPhysicL::raw_value(y) == 0 ? abs_tol_check : rel_tol_check;
```
But my desire is that we use a relative convergence check when we can

joshuahansel replied:
I know, but supposing y can be any value (and I haven't thought about if this would ever/often be the case), you'll always fail to converge if y is very small but not zero. For example, supposing your tolerance is 1e-8 and y is 1e-12, you'd be requiring that f be < 1e-20.

lindsayad replied:
that's a good point

lindsayad committed:
Remove jacobian checking section

Function perturbation doesn't appear to work for tables?

lindsayad committed:
Use two D Newton inversion function signature closer to original

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1240570/) on 91a09b8 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>3b891b</th>
      <th colspan=4>#22739 91a09b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.39%</td>
      <td>84.39%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80649</td>
      <td>80652</td>
      <td>+3</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14916</td>
      <td>14913</td>
      <td>-3</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22739/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22739/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>3b891b</th>
      <th colspan=4>#22739 91a09b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.71%</td>
      <td>83.10%</td>
      <td>-1.61%</td>
      <td>79.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6155</td>
      <td>6108</td>
      <td>-47</td>
      <td>678</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1111</td>
      <td>1242</td>
      <td>+131</td>
      <td>173</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22739/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22739/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22739/coverage/xfem)

</details>

## Warnings

- `fluid_properties` new line coverage rate 79.67% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1240570/) on 91a09b8 wanted to post the following:


The following coverage requirement(s) failed:
- `fluid_properties` coverage rate 83.10% is less than the required 84.0%

lindsayad committed:
Increase Water97FluidProperties unit testing

lindsayad committed:
Address Josh review

- Use a tolerance when deciding whether to do an absolute or relative
  convergence check
- Use 'R' instead of 'f' in function to avoid confusion with doxygen
- Rename `T_from_p_rho` to `T_drhodT_from_p_rho`

lindsayad committed:
Create THM Water97FluidProperties jacobian test

GiudGiud commented on modules/fluid_properties/include/userobjects/SinglePhaseFluidProperties.h:
```diff
@@ -405,3 +415,67 @@ class SinglePhaseFluidProperties : public FluidProperties
+}
+
+template <>
+inline std::pair<Real, Real>
```
why the instantiation?

lindsayad replied:
Because you cannot construct a Real with the same constructor arguments as you can dual numbers

GiudGiud commented on modules/fluid_properties/include/utils/NewtonInversion.h:
```diff
@@ -19,18 +20,73 @@ namespace FluidPropertiesUtils
+    // Check the accuracy of the Jacobian
+    auto J_differenced = (perturbed_y - new_y) / (1e-8 * z);
+    if (!MooseUtils::relativeFuzzyEqual(J_differenced, dy_dz, 1e-2))
+      mooseError("Bad Jacobian in NewtonSolve");
```
```suggestion
      mooseWarning("Bad Jacobian in NewtonSolve");
```

some derivatives are implemented by finite difference, so we could hit this in acceptable conditions

joshuahansel commented:
> > It seems odd that we try to invert negative energies and enthalpies though. I d double check those grids to see why one of the bounds is negative
> 
> The csv file has negative internal energy so that's what being written out by @cpgr's original test. I guess since internal energy includes a potential which you can kind of translate arbitrarily, it can be negative ... ?

I'd argue that {e,h,s} are fine to be negative, so long as they are consistent. They all have some reference state that is defined as the zero value.

lindsayad committed:
More generification of SinglePhaseFluidProperties

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1241443/) on a766e2c wanted to post the following:

The following commit(s) contain fixups:

- c01629b7b85: fixup! Increase Water97FluidProperties unit testing

lindsayad committed:
Update Stiffened and CaloricallyImperfect fp classes

lindsayad committed:
Don't implement brittle default implementations

lindsayad committed:
Finish implementing s_from_p_T in Water97FluidProperties

E.g. implement the five function parameter template

lindsayad committed:
Fix PetscInfo for Petsc version >= 3.17

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1242556/) on eaafe26 wanted to post the following:


The following coverage requirement(s) failed:
- `fluid_properties` coverage rate 83.92% is less than the required 84.0%

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/1243649/) on 3245cfd : invalidated by @lindsayad

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1243650/) on 3245cfd : invalidated by @lindsayad

lindsayad commented:
This PR definitely shows the effects of the `-O2` compilation on modules coverage @loganharbour. Small templates are definitely getting inlined and not showing up in coverage

lindsayad committed:
Expand testing of StiffenedGasFluidProperties

licharlot commented on modules/fluid_properties/include/utils/NewtonInversion.h:
```diff
@@ -19,18 +20,73 @@ namespace FluidPropertiesUtils
+
+  // Check for divergence or slow convergence of Newton's method
+  if (iteration >= max_its)
+    mooseError(
```
Could you make this a `mooseException` instead? This is stopping simulations that would otherwise go to completion (with the tabulated FP).

lindsayad committed:
errors to exceptions in FP Newton inversion

joshuahansel commented on modules/thermal_hydraulics/test/tests/components/inlet_mass_flow_rate_1phase/jac.massflowrate_3eqn_water97.i:
```diff
@@ -0,0 +1,87 @@
+[GlobalParams]
```
What does this test test that the existing one does not, in regards to `InletMassFlowRateTemperature1Phase`?

licharlot replied:
We need to figure out a way to test the fluid property compatibility with THM. For example, there is a test for the calorically imperfect gas in the VolumeJunction1Phase tests because the lookups in that component were causing problems at some point.

GiudGiud replied:
Maybe we should make a simple yet maximal input file for THM and NS and place in a special test folder like navier_stokes/test/tests/fluid_prop_compatibility (same for thm). 

Then every time we add a FP we add an inactive (only active for one test) block in that input file.

joshuahansel replied:
Yeah, we could do something like that. I won't require it in this PR.

As for the tagging system, that sounds nice, but may be a lot of work, and has the potential to get complicated (like if there are multiple flow models in THM/NS).

lindsayad replied:
Yes I just wanted to be able to run a THM problem with `Water97FluidProperties` ... I did not realize I would only get to that point almost 3k lines later :grimacing: :laughing:

lindsayad replied:
I personally think that most fluid properties classes should go in the module that uses them ... then at least you would instantly know one module that the FP object is compatible with ... and then it's a little bit easier to get coverage testing as well. But I also agree that a table in documentation would be great

GiudGiud replied:
I ll add the table soon then

joshuahansel replied:
I like to just run it and see if it works - it makes it more exciting. Also, that list could be a little restrictive if for example, some obscure interface is used by some component that is rarely used. Maybe if you're going to make a table, just mention that possibility - that they can just try it, even though the table says something else.

joshuahansel commented on modules/fluid_properties/unit/src/Water97FluidPropertiesTest.C:
```diff
@@ -677,3 +677,97 @@ TEST_F(Water97FluidPropertiesTest, henry)
+    auto e = _fp->e_from_p_T(pressure, temperature);
+
+    auto [p_test, T_test] = _fp->p_T_from_v_e(v, e);
+    REL_TEST(MetaPhysicL::raw_value(pressure), MetaPhysicL::raw_value(p_test), REL_TOL_CONSISTENCY);
```
You put all of the `raw_value` in `REL_TEST` and `ABS_TEST`, didn't you?

lindsayad replied:
Yes I did do that "later". I can remove the redundancy

joshuahansel commented on modules/fluid_properties/test/include/utils/SinglePhaseFluidPropertiesTestUtils.h:
```diff
@@ -28,12 +28,32 @@
+    ADReal f_value, df_da, df_db;                                                                  \
+    f(a, b, f_value, df_da, df_db);                                                                \
+    REL_TEST(                                                                                      \
+        MetaPhysicL::raw_value(f(a, b)), MetaPhysicL::raw_value(f_value), REL_TOL_CONSISTENCY);    \
```
Same as my other comment - you've already put `raw_value` in `REL_TEST`

joshuahansel commented on modules/fluid_properties/include/utils/NewtonInversion.h:
```diff
@@ -42,21 +98,131 @@ Real NewtonSolve(const Real & x,
+  // Check for divergence or slow convergence of Newton's method
+  if (iteration >= max_its)
+    mooseException(
+        "Newton solve convergence failed: maximum number of iterations, ", max_its, " exceeded");
```
, exceeded

joshuahansel commented on modules/fluid_properties/include/utils/NewtonInversion.h:
```diff
@@ -19,18 +20,73 @@ namespace FluidPropertiesUtils
+  // Check for divergence or slow convergence of Newton's method
+  if (iteration >= max_its)
+    mooseException(
+        "Newton solve convergence failed: maximum number of iterations, ", max_its, " exceeded");
```
, exceeded

lindsayad committed:
Address Josh review

- fix typos
- don't do redundant `raw_value`
- matching commas

moosebuild commented:
Job [Conda (Ubuntu)](https://civet.inl.gov/job/1246199/) on 8b2b397 : invalidated by @lindsayad

GiudGiud commented on modules/fluid_properties/include/userobjects/StiffenedGasFluidProperties.h:
```diff
@@ -115,3 +107,167 @@ class StiffenedGasFluidProperties : public SinglePhaseFluidProperties, public Na
 
 #pragma GCC diagnostic pop
+
+template <typename CppType>
```
same for the calorifically imperfect gas

lindsayad replied:
I think that if we imagine that these are the final versions of classes, then I can mark them as `final`, make these templates private and then move to them to the `.C`. What do you think?

GiudGiud replied:
I think that would be fine.
A user could still override the routine calling the template

GiudGiud replied:
well "annoying" is an overstatement.

We do out of a deficiency of C++ compiling of templates though. We generally prefer source in source files?

lindsayad replied:
I do not have much preference. If a non-template is a good candidate for inlining then I like it in the header with `inline`. I don't mind having templates in headers whenever they may be used by other translation units. The compilation speed should be a function of number of instantiations irrespective of whether the implementation is in header or .C I *think* (not totally sure). Anyway I'll put this into the .C. If another TU ever wants to use the templates then I can move them and make them `protected`

GiudGiud commented on modules/fluid_properties/include/userobjects/Water97FluidProperties.h:
```diff
@@ -57,24 +58,72 @@ class Water97FluidProperties : public SinglePhaseFluidProperties
 
   virtual void
   e_from_p_T(Real pressure, Real temperature, Real & e, Real & de_dp, Real & de_dT) const override;
+  virtual void e_from_p_T(const ADReal & pressure,
```
is this really needed?

lindsayad replied:
you prefer more stack frames?

GiudGiud replied:
asking because this is the first case I see of this outside of rho_from_p_T
looks like there's also e_from_p_rho and h_from_p_T I ve never seen in that form as well

lindsayad replied:
And these implementations are necessary in order to get derivatives correct for newton inversions

lindsayad replied:
You have to be able to compute `ADReal dy_dx` in order get the `ADReal` from newton inversion to be accurate. A `Real dy_dx` won't get it done

GiudGiud replied:
ah that's the use case

lindsayad commented:
Yea and then making those routines public again results in undefined references in the unit testing translation unit because the public methods use private methods defined in the `Water97FluidProperties.C` translation unit. This is why I put templates in the header to avoid this. This is a waste of time

GiudGiud commented:
kk let s ditch the commit then

I didnt know that unit testing would trigger this, sorry about that
